### PR TITLE
Implement cupy.linalg.tensorinv

### PR DESCRIPTION
**Pull request by rezoo**
This is a re-submission of #462 (I closed this PR by mistake).
This PR implements `cupy.linalg.tensorinv`.

### CODE REVIEW

**Comment by rezoo** [comment]
> Cloud you check #454?
I see, but in this case the value of `ind` depends on the shape, and therefore it would be hard to apply `testing.product` directly.
And it seems that #454 does not check the equivalence of arrays. Is it OK?

**Comment by okuta** [comment]
We can use `testing.parameterize` independently.

>And it seems that #454 does not check the equivalence of the types of arrays. Is it OK?
OK, we need more test.

**Comment by rezoo** [comment]
This means that I should use `testing.parameterize`, right?

**Comment by okuta** [comment]
Yes.

File: cupy/linalg/solve.py [hunk 1]
```diff
@@ -192,4 +192,33 @@ def pinv(a, rcond=1e-15):
...
+    Returns:
+        cupy.ndarray:
+            The inverse of a tensor whose shape is equivalent to
+            ``a.shape[ind:] + a.shape[:ind]``.
+
+    .. seealso:: :func:`numpy.linalg.tensorinv`
+    '''
+    util._assert_cupy_array(a)
+
+    if ind <= 0:
```
Comment (unnonouno) [inline cupy/linalg/solve.py]: Please add a test for this case.

**Comment by unnonouno** [comment]
jenkins test this please

File: cupy/linalg/solve.py (hunk 1 above)
Comment (rezoo) [reply cupy/linalg/solve.py]: Done.

**Comment by rezoo** [comment]
After discussing with okuta-san, I decided to move `retry` decorators instead of `testing.parameterize` for efficiency.

**Comment by rezoo** [comment]
Sorry for being late. Please check the updates.

**Comment by unnonouno** [comment]
Why not using parameterized test? Usually it is better that one test method has one test, because when an error occurs, developers easily understand its error condition.

**Comment by rezoo** [comment]
I internally discussed with okuta-san, and claimed that `@testing.parameterize` does not fit with the current situation:

1. We cannot use `product` because parameters are not independent.
2. We have to split this into two classes if we want to use `@testing.parameterize`, however, the number of lines of code increases.

In addition, I am not convinced in your claim; if it fails, developers can easily understand the condition even in the current test code because it outputs the point of failure. Is this argument wrong? It seems that in this case readability gets slightly worse due to the explicit use of decorator.

**Comment by unnonouno** [comment]
Hum, OK. LGTM

**Pull request merged**